### ISSUE TITLE
[PHP] Cookies are randomly corrupted

### ISSUE THREAD

**Issue by Peter2121** [nginx/unit#842]
Unit 1.29.0 with PHP 8.2.3 in FreeBSD 13.1 jail.
Cookies set by PHP setcookie() are randomly corrupted after sending form (with the form data and zeros). If session_start() is used - PHPSESSID cookie is corrupted, session data is lost.

Code to reproduce the issue:

```
<?php
session_start();

if(!count($_COOKIE))
    $_SESSION['sessioninit'] = '1';

if(isset($_POST['public_form']))
    $_SESSION['form_sended'] = true;

echo '<pre>';
echo '<h1>Liste des cookies</h1>';
        print_r($_COOKIE);
echo '<h1>Infos session</h1>';
        print_r($_SESSION);
echo '<h1>Session Id</h1>';
                echo session_id();

session_write_close();

        echo '<br/><br/><form method="POST"><input type="hidden" name="public_form" value="sended">'
        . '<input id="public_form_submit" class="button submit" type="submit" value="Valider">'
        . '</div></form>';
        die;
```

Reloaded page (after cookie set) chows correct cookie:

`[PHPSESSID] => bgf6gjmi7te0vfmst6fjd3m8ds`

...and `sessioninit` data.

Sending form (with 'Valider' button) shows corrupter cookie:

`[PHPSESSID] => bgf6gjmi7te0vfmst6fjd3m8dspublic_form=sended���...`

Session data `sessioninit` is lost.

The same code works correctly with PHP-FPM and with Unit 1.28 and PHP 7.4.

If multiple cookies are set - a randomly selected one is corrupted at the same manner.

**Comment by ac000** [comment]
The above test case seems to work OK on Linux (Fedora 37) with Unit 1.29.0 and PHP 8.1.15 and FreeBSD 13.1 with PHP 8.2.0 installed via 'pkg install lang/php82 php82-session'

What does your unit config look like and how did you set up the jail on FreeBSD? Perhaps you could test without the jail?, doubtful it will make any difference, but you never know...

How did you install PHP 8.2.3?

**Comment by Peter2121** [comment]
Voila my config files. All software is installed from packages ('latest' repository). I had the same error on PHP 8.2.0 before upgrade of packages. Jail is managed by CBSD (readonly base). Maybe it is important that Unit is behind NGINX (mode proxy using unix socket) - you can see it in my config files.
Sorry, I have no VM to test without jail.

[nginx.conf.txt](https://github.com/nginx/unit/files/10859189/nginx.conf.txt)
[unit_backend.conf.txt](https://github.com/nginx/unit/files/10859190/unit_backend.conf.txt)
[php.ini.txt](https://github.com/nginx/unit/files/10859191/php.ini.txt)
[unit-config.json.txt](https://github.com/nginx/unit/files/10859192/unit-config.json.txt)

**Comment by Peter2121** [comment]
NGINX version is 1.22.1

**Comment by ac000** [comment]
OK.

Could you try disabling compression in nginx?

If you still see the problem after that, could you try _without_ nginx, just hitting unit directly?

**Comment by Peter2121** [comment]
Trying to reconfigure Unit/NGINX to hit Unit directly, I changed unix socket to tcp socket. The problem has disappeared! So the problem is related to unix socket, used by Unit.
We use unix sockets to simplify the configuration of Unit so it does not depend on IP addresses of jails. Please, fix this bug.

**Comment by ac000** [comment]
So to confirm. You changed unit to listen on a TCP socket but still using nginx?

**Comment by Peter2121** [comment]
Yes, exactly

**Comment by ac000** [comment]
OK, I managed to reproduce this using nginx as a proxy. Reproduced on both FreeBSD (as above) and Linux (Fedora 37 with current unit from git and PHP 8.1.15).

It seems to be an issue with TLS and Unix domain sockets, doing plain http to nginx seems to work fine talking to unit over a Unix domain socket.

More investigation required.

**Comment by ac000** [comment]
So the problem is really just down to Unix domain sockets. Things just _seemed_ to work OK without TLS to nginx and UDS  to unit, but things are actually not OK in that case either.

When doing TLS to nginx and UDS from nginx to unit, looking at the memory layout in GDB we see

```gdb
(gdb) x /64bs 0x7f6141f3448a-768
...
0x7f6141f34423: "HTTP_SEC_FETCH_SITE"
0x7f6141f34437: "same-origin"
0x7f6141f34443: "HTTP_SEC_FETCH_USER"
0x7f6141f34457: "?1"
0x7f6141f3445a: "HTTP_COOKIE"
0x7f6141f34466: "PHPSESSID=uo5b2nu9buijkc89jotbgmd60vpublic_form=sended"
0x7f6141f3449d: ""
```

This is after unit has tried to create a bunch of header=value pairs. As you can see, the form data has been appended to the cookie value.

Lets look at the same again, but without TLS

```gdb
(gdb) x /64bs 0x7f6141f343d2-512
...
0x7f6141f3435e: "HTTP_REFERER"
0x7f6141f3436b: "http://fedora/842.php"
0x7f6141f34381: "HTTP_COOKIE"
0x7f6141f3438d: "PHPSESSID=a86egdu2ak6tbo3tsbffo8chss"
0x7f6141f343b2: "HTTP_UPGRADE_INSECURE_REQUESTS"
0x7f6141f343d1: "1public_form=sended", '\245' <repeats 184 times>
0x7f6141f3449d: ""
```

Again you can see that the form data has been appended to the last header value. The reason this seems to work OK, is that in this case the cookie is not the last value. The other reason this works is

```gdb
(gdb) p req->content_buf->free
$3 = 0x7f6141f343d2 "public_form=sended", '\245' <repeats 184 times>
```

When using a normal TCP socket between nginx and unit we get

```gdb
(gdb) x /64bs 0x7f6141f3448a-512
...
0x7f6141f34432: "HTTP_SEC_FETCH_USER"
0x7f6141f34446: "?1"
0x7f6141f34449: "HTTP_COOKIE"
0x7f6141f34455: "PHPSESSID=uo5b2nu9buijkc89jotbgmd60v"
0x7f6141f3447a: "\245\245\245\245\245\245\245\245\245\245public_form=sended\245\245\245\245\245\245"
0x7f6141f3449d: ""
```

We can see that the form data is clearly separated from the headers.

It's still not clear why this is happening. The data received by unit when using a UDS or a normal socket seems identical

UDS

```strace
recvfrom(27<UNIX-STREAM:[82386->83885,"/dev/shm/842.sock"]>, "
POST /842.php HTTP/1.1\r\n
Host: fedora\r\n
X-Forwarded-Proto: https\r\n
X-Real-IP: 2001:db8::f00\r\n
X-Forwarded-For: 2001:db8::f00\r\n
Content-Length: 18\r\n
user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/110.0\r\n
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\n
accept-language: en-GB,en;q=0.5\r\n
accept-encoding: gzip, deflate, br\r\n
content-type: application/x-www-form-urlencoded\r\n
origin: https://fedora\r\n
referer: https://fedora/842.php\r\n
upgrade-insecure-requests: 1\r\n
sec-fetch-dest: document\r\n
sec-fetch-mode: navigate\r\n
sec-fetch-site: same-origin\r\n
sec-fetch-user: ?1\r\n
cookie: PHPSESSID=uap24vclgf3kom6cq7q75fcrcs\r\n
\r\n
public_form=sended
", 2048, 0, NULL, NULL) = 709
```

TCP socket

```strace
recvfrom(14<TCPv6:[[::1]:8080->[::1]:47928]>, "
POST /842.php HTTP/1.1\r\n
Host: fedora\r\n
X-Forwarded-Proto: https\r\n
X-Real-IP: 2001:db8::f00\r\n
X-Forwarded-For: 2001:db8::f00\r\n
Content-Length: 18\r\n
user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/110.0\r\n
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\n
accept-language: en-GB,en;q=0.5\r\n
accept-encoding: gzip, deflate, br\r\n
content-type: application/x-www-form-urlencoded\r\n
origin: https://fedora\r\n
referer: https://fedora/842.php\r\n
upgrade-insecure-requests: 1\r\n
sec-fetch-dest: document\r\n
sec-fetch-mode: navigate\r\n
sec-fetch-site: same-origin\r\n
sec-fetch-user: ?1\r\n
cookie: PHPSESSID=uap24vclgf3kom6cq7q75fcrcs\r\n
\r\n
public_form=sended
", 2048, 0, NULL, NULL) = 709
```

Still investigating...

**Comment by ac000** [comment]
Having done a ```git-bisect(1)```  between 1.28.0 & 1.29.0, it _correctly_ pointed to an earlier commit, a03274456b54cbc39e220b9dd73c3fc3fb935e46, of mine as the cause of the issue.

This would also affect normal TCP sockets, except we seem to allocate a little more than we need in that case, which means we avoid this issue there.

Feel free to test the following patch, it fixes it for me (tested on Linux and FreeBSD).

```diff
diff --git a/src/nxt_router.c b/src/nxt_router.c
index 2b49df0d..7ab221a6 100644
--- a/src/nxt_router.c
+++ b/src/nxt_router.c
@@ -5243,6 +5243,7 @@ nxt_router_prepare_msg(nxt_task_t *task, nxt_http_request_t *r,
                + r->version.length + 1
                + r->remote->length + 1
                + r->local->length + 1
+               + nxt_sockaddr_port_length(r->local) + 1
                + r->server_name.length + 1
                + r->target.length + 1
                + (r->path->start != r->target.start ? r->path->length + 1 : 0);
```

**Comment by Peter2121** [comment]
The first tests show that the problem has disappeared.
I ask my colleagues to test a little bit more.
Thank you!!

**Comment by ac000** [comment]
This has been merged in 78e1122a3c94a150219b4b6e1e594ae5bfdd8d68
